lib/Dialect/Arc/Transforms/LowerState.cpp

```diff
@@ -619,70 +626,77 @@ LogicalResult ModuleLowering::lowerExtModule(InstanceOp instOp) {
     baseName.resize(baseNameLen);
     baseName += '/';
     baseName += cast<StringAttr>(name).getValue();
-    auto state = builder.create<AllocStateOp>(
+    auto state = stateBuilder.create<AllocStateOp>(
         result.getLoc(), StateType::get(intType), storageArg);
-    state->setAttr("name", builder.getStringAttr(baseName));
+    state->setAttr("name", stateBuilder.getStringAttr(baseName));
     replaceValueWithStateRead(result, state);
   }
 
-  if (&*builder.getInsertionPoint() == instOp)
-    builder.setInsertionPointAfter(instOp);
   instOp.erase();
   return success();
 }
 
 LogicalResult ModuleLowering::cleanup() {
+  // Clean up dead ops in the model.
+  SetVector<Operation *> erasureWorklist;
+  auto isDead = [](Operation *op) {
+    if (isOpTriviallyDead(op))
+      return true;
+    if (!op->use_empty())
+      return false;
+    if (auto stateOp = dyn_cast<StateOp>(op))
+      return stateOp.getLatency() == 0;
```

Comment:
Isn't it a problem if we delete state ops that still have uses? If there are no uses it should already be covered above.

Reply:
That's what the `if` statements above do. If the op is trivially dead, return true. If the op wasn't trivially dead because it has uses (and not because of side effects), return false. Otherwise check if this is a lat=0 state ops, which has side effects and isn't covered by `isOpTriviallyDead`.